etl/steps/data/meadow/biodiversity/2024-01-25/cherry_blossom.py

```diff
@@ -0,0 +1,89 @@
+from datetime import datetime
+from typing import cast
+
+from owid.catalog import Table
+from owid.catalog import processing as pr
+from structlog import get_logger
+
+from etl.helpers import PathFinder, create_dataset
+from etl.snapshot import Snapshot
+
+log = get_logger()
+
+# naming conventions
+paths = PathFinder(__file__)
+
+
+def run(dest_dir: str) -> None:
+    log.info("cherry_blossom.start")
+
+    # retrieve snapshot
+    snap = cast(Snapshot, paths.load_dependency("cherry_blossom.xls"))
+    tb = snap.read(skiprows=25)
+
+    # clean and transform data
+    tb = clean_data(tb)
+    tb = add_data_for_recent_years(tb)
+    tb = convert_date(tb)
+    tb = tb.set_index(["country", "year"], verify_integrity=True)
+    #
+    # Save outputs.
+    #
+    # Create a new grapher dataset with the same metadata as the garden dataset.
+    ds = create_dataset(dest_dir, tables=[tb], default_metadata=snap.metadata)
+
+    # finally save the dataset
+    ds.save()
+
+    log.info("cherry_blossom.end")
+
+
+def clean_data(tb: Table) -> Table:
+    tb = tb.dropna(subset=["Full-flowering date (DOY)"])
+    tb["country"] = "Japan"
+    tb = tb.rename(columns={"AD": "year"}).drop(columns=["Source code", "Data type code", "Reference Name"])
+    return tb
+
+
+def convert_date(tb: Table) -> Table:
+    """
+    The full flowering date is formated like MDD, we should change this to day of the year for better biological meaning. For example the 4th April is shown as 404.
+    In this function we:
+     - Zero pad the year so it is 4 digits long - the data starts in 812
+     - Zero pad the full-flowering date so it is 4 digits long
+     - Combine these year and the full-flowering date so 1st April 812 would look like 08120401
+     - Convert this into a date format and extract the Julian day (day of the year)
+    """
+    tb["year_zpad"] = tb["year"].astype("str")
+    tb["year_zpad"] = tb["year_zpad"].str.zfill(4)
+    tb["Full-flowering date"] = tb["Full-flowering date"].astype(float).astype("Int64").astype("str").str.zfill(4)
+    tb["date_combine"] = tb["year_zpad"] + tb["Full-flowering date"]
+
+    tb["Full-flowering date"] = tb["date_combine"].apply(lambda x: datetime.strptime(x, "%Y%m%d").strftime("%j"))
+
+    tb = tb.drop(columns=["Full-flowering date (DOY)", "date_combine", "year_zpad"])
+
+    return tb
+
+
+def add_data_for_recent_years(tb: Table) -> Table:
+    """
+    Not sure if this is the best way to do this...
+    Adding data for the years 2016-21, from here:
```

Comment:
I suppose that ideally you want to create a snapshot step with this data? 

The snapshot step would just be similar to this function, where you create the table manually (retrieving from the site seems complex). You can specify how the transcription is done from the site to the table in the code.

The good part of having it in snapshot is that you can capture the metadata from the source in the dvc file next to the script.

Reply:
Yeh that's a good point, I wasn't sure best how to tackle it. I was slightly hesitant to create another snapshot as all the data is basically from the same producer, and I was worried it would cause a lot of metadata duplication. Do you think it would be okay?

Reply:
If I understand correctly, the same data provider is publishing the data until 2015 in a file (http://atmenv.envi.osakafu-u.ac.jp/omu-content/uploads/sites/1215/2015/10/KyotoFullFlower7.xls), and then 2016-2021 online on their site? Plus, 2021-2023 is obtained via personal correspondence.

This is not super important, but the user could be confused by looking at the sources and the chart and seeing that the year covered by both differs.

You could still have a snapshot file for 2016-2023, explaining that this is obtained from the online site and complemented with data provided via private communication. Then, two sources would be listed in the sources tab: the Excel file and one without a link representing the manually inputted data.

I may be lacking some context here, but we could seduce the author to release everything in one file at some point in the future.